soco/events_base.py

```diff
@@ -0,0 +1,642 @@
+        if requested_timeout is not None:
+            headers["TIMEOUT"] = "Second-{}".format(requested_timeout)
+
+        # pylint: disable=missing-docstring
+        def success(headers):
+            timeout = headers['timeout']
+            # According to the spec, timeout can be "infinite" or "second-123"
+            # where 123 is a number of seconds.  Sonos uses "Second-123"
+            # (with a capital letter)
+            if timeout.lower() == 'infinite':
+                self.timeout = None
+            else:
+                self.timeout = int(timeout.lstrip('Second-'))
+            self._timestamp = time.time()
+            self.is_subscribed = True
+            log.info(
+                "Renewed subscription to %s, sid: %s",
+                self.service.base_url + self.service.event_subscription_url,
+                self.sid)
+
+        # pylint: disable=missing-docstring
+        def failure():
+
+            # If the renewal has failed, it seems appropriate
+            # to cancel the subscription.
+            self._cancel_subscription()
+
+            raise SoCoException(
+                "Could not renew subscription to {}, sid: {}".format(
+                    self.service.base_url +
+                    self.service.event_subscription_url,
+                    self.sid))
+
+        self.request(
+            'SUBSCRIBE',
+            self.service.base_url + self.service.event_subscription_url,
+            headers, success, failure)
+
+    def unsubscribe(self):
+        """unsubscribe()
+        Unsubscribe from the service's events.
+        Once unsubscribed, a Subscription instance should not be reused
+        """
+        # Trying to unsubscribe if already unsubscribed, or not yet
+        # subscribed, fails silently
+        if self._has_been_unsubscribed or not self.is_subscribed:
+            return
+        self._cancel_subscription()
+
+        # If the subscription has timed out, an attempt to
+        # unsubscribe from it will fail.
+        if self.time_left == 0:
+            log.info(
+                "Subscription %s, sid: %s has expired. No need to unsubscribe",
+                self.service.base_url + self.service.event_subscription_url,
+                self.sid)
+            return
+
+        # Send an unsubscribe request like this:
+        # UNSUBSCRIBE publisher path HTTP/1.1
+        # HOST: publisher host:publisher port
+        # SID: uuid:subscription UUID
+        headers = {
+            'SID': self.sid
+        }
+
+        # pylint: disable=missing-docstring, unused-argument
+        def success(*arg):
+            log.info(
+                "Unsubscribed from %s, sid: %s",
+                self.service.base_url + self.service.event_subscription_url,
+                self.sid)
+
+        # pylint: disable=missing-docstring
+        def failure():
+            log.warning(
+                "Error attempting to unsubscribe from %s, sid: %s",
```

Comment:
I'm trying to remember the different circumstances there have been unsubscribe failures. 

One was on attempts to unsubscribe when a subscription has timed out. This has been addressed, hopefully, by not attempting to unsubscribe if time_left is 0.

Another, I think, was (randomly?) when unsubscribe was invoked via atexit, but I'm not 100% sure.

I'll aim to do some more testing and see if I can replicate the failures. That should give us a better idea how best to deal with them.

Reply:
I've tested this further. The only scenarios in which I've been able to produce errors upon unsubscribe or auto_renew is upon network disconnects, now that the timeout problem has been addressed. 

Raising an exception on auto_renew seems pointless as there's no calling code to catch the error. Therefore I think it's appropriate to log a warning.

Raising an exception on subscribe does seem right, as the subscribe attempt will have failed and the calling code needs to deal with that.

On reflection, I think a warning is appropriate on unsubscribe for the following reasons:

- It's not quite true to say the unsubscribe attempt has failed (in the way a subscribe attempt would have failed). The subscription has been cancelled and deleted from the list of subscriptions. In that sense, it has been unsubscribed. It's just that the notification of 'UNSUBSCRIBE' to the sonos system failed, most likely because the network or the sonos system was down. It's hard to see what more the calling code could do about that, beyond what has already been done. It seems appropriate to warn about the problem through the logging system, rather than raise an exception.

- Unsubscribe is called automatically at exit / shutdown, where there is again no calling code to catch the exception. It also seems these exceptions at exit aren't reported by python. Query whether that means unsubscribe isn't being fully executed at exit.

My proposal therefore is to raise an exception on a subscribe failure, but to log a warning on an auto_renew or unsubscribe failure.

Reply:
Thanks for the research and clarification. Now I agree that subscribe should raise an error and unsubscribe should log a warning.
Originally, I misunderstood the code and thought that renew() would just log an error as well, but I see now that it throws an error. I think this is right because we can't take for granted that it will only ever be called by the auto renew thread.

While we're at it: What happens if the auto renewal fails? Currently, the exception remains uncaught, so the exception is written to stderr by the interpreter.

An alternative would be this (in `AutoRenewThread.run()`):
```python
def run(self):
    subscription = self.subscription
    stop_flag = self.stop_flag
    interval = self.interval
    try:
        while not stop_flag.wait(interval):
            log.info("Autorenewing subscription %s", sub.sid)
            subscription.renew()
    except SoCoException as exc:
        log.error(str(exc))

Reply:
Your exception handling is good as well, the raise-catch-log pattern isn't everyone's taste.
But if you want to implement my version, here's how I would do it with twisted:
```python
    def _auto_renew_start(self, interval):
        """Starts the auto_renew loop."""
        def try_renew():
            log.info("Autorenewing subscription %s", self.sid)
            try:
                self.renew()
            except SoCoException as exc:
                log.error(str(exc))
                self._auto_renew_loop.stop()
        self._auto_renew_loop = task.LoopingCall(try_renew)
        # False means wait for the interval to elapse, rather than fire at once
        self._auto_renew_loop.start(interval, False)

Reply:
Thank you! Twisted has a different approach to Exception handling. I think that must have been what drove my initial approach of logging errors rather than raising them. On reflection, as your comments show, that's not the right approach. I'm going to work through this next week to come up with something that handles errors better. 

In the Subscription class in events_twisted, the twisted method that sends HTTP requests (used by the subscribe, unsubscribe and renew methods) returns a [Deferred](https://twistedmatrix.com/documents/current/core/howto/defer-intro.html). If an Exception occurs inside a function that returns a Deferred, a [Failure](http://twistedmatrix.com/documents/current/api/twisted.python.failure.Failure.html) object is passed to the first errback that has been registered for that deferred with [Deferred.addErrback](https://twistedmatrix.com/documents/current/api/twisted.internet.defer.Deferred.addErrback.html).

I think the subscribe, unsubscribe and renew methods need to return a Deferred in the events_twisted version and to return None in the events version. Then, in the twisted version, the calling code can use addErrback to deal with failures.

Additionally, there should probably be a 'lenient' flag that, if set, will cause the subscribe, unsubscribe and renew methods to log Exceptions rather than raise them. The lenient flag will be used by internal calls. Autorenew is the example we have been discussing, but there's also automatic unsubscribe at exit.

I haven't yet thought through a solution to service.subscribe. This returns a Subscription object, after first calling its subscribe method. In the non-twisted version, if an Exception occurred during the HTTP request call in Subscription.subscribe, it could be caught and handled by the upstream code that called service.subscribe. 

In the twisted version, there would instead be an unhandled failure. The calling code receives a Subscription object, not a Deferred. Therefore, the calling code cannot register an errback to receive and handle the failure. One possibility would be to set a property of the Subscription object to refer to such a failure, so that the calling code could check that property after calling service.subscribe().

Reply:
This is really interesting. The subscribe() and unsubscribe() methods return immediately, so afterwards 1) the subscription won't be finished yet and 2) you don't know if an exception will occur.
The twisted way is to return a Deferred and do 1) addCallback for when the subscription has finished and 2) addErrback for when it has thrown an exception.

Now the question is, what should service.subscribe do?

- Just return the Subscription (current behavior): User can't react to events. :x:
- Just return the Deferred: User doesn't get the Subscription. :x:
- Return a (Subscription, Deferred) tuple: Changes the method signature. :x:
- Return a Subscription with the Deferred as an attribute: Possible, but unintuitive (and probably cumbersome to implement) :white_check_mark:
- Return a Deferred with the Subscription as an attribute: Possible, but unintuitive :white_check_mark: 
- Return a Deferred which passes the Subscription to the callback: Intuitive, but user gets the Subscription only after subscribe call has returned :white_check_mark:
- Return a Deferred which passes the Subscription to the callback, and has the Subscription as an attribute: Like the above, but the user can get the Subscription directly if he wants to :heavy_check_mark:

So my favored option is the last one.

Reply:
Some code:
```python
# events_base.SubscriptionBase
def subscribe(self, ...):
    ...

    def success(headers):
        ...

    with self.subscriptions.subscriptions_lock:
        return self._request('SUBSCRIBE', service.base_url + service.event_subscription_url,
            headers, success)
```

```python
# events.Subscription
def _request(self, method, url, headers, success=None, failure=None):
    try:
        response = requests.request(method, url, headers=headers)
        response.raise_for_status()
    except requests.exceptions.RequestException as exc:
        if failure:
            failure(exc)
    else:
        if success:
            success(response.headers)
        return self
```

```python
# events_twisted.Subscription
def _request(self, method, url, headers, success=None, failure=None):
    ...
    d = agent.request(*args)

    if success:
        def on_success(response):
            ...
            success(response_headers)
            return self
        d.addCallback(on_success)
    
    if failure:
        def on_failure(fail):
            failure(fail.value)
        d.addErrback(on_failure)

    d.subscription = self
    return d
```

```python
# services.Service
def subscribe(self, ...):
    subscription = config.EVENTS_MODULE.Subscription(
            self, event_queue)
    return subscription.subscribe(...)
```

Reply:
Looks good!